crates/sui-types/src/error.rs

```diff
@@ -417,3 +433,193 @@ impl ExecutionStateError for SuiError {
+    ModuleNotFound,
+    /// #[error("Function signature is invalid: {error:?}.")]
+    InvalidFunctionSignature,
+    /// #[error("Function visibility is invalid for an entry point to execution: {error:?}.")]
+    InvalidFunctionVisibility,
+    InvalidNonEntryFunction,
+    ExecutionInvariantViolation,
+    /// #[error("Type error while binding function arguments: {error:?}.")]
+    TypeError,
+    /// Circular object ownership detected
+    CircularObjectOwnership,
+    MissingObjectOwner,
+    InvalidSharedChildUse,
+
+    ModulePublishFailure,
+    ModuleVerificationFailure,
+
+    // Move Error
+    VmError,
+}
+
+impl std::fmt::Display for ExecutionErrorKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ExecutionErrorKind: {:?}", self)
+    }
+}
+
+impl std::error::Error for ExecutionErrorKind {}
+
+#[derive(Debug)]
+pub struct ExecutionError {
+    inner: Box<ExecutionErrorInner>,
+}
+
+#[derive(Debug)]
+struct ExecutionErrorInner {
+    kind: ExecutionErrorKind,
+    source: Option<BoxError>,
+}
+
+impl ExecutionError {
+    pub fn new(kind: ExecutionErrorKind, source: Option<BoxError>) -> Self {
+        Self {
+            inner: Box::new(ExecutionErrorInner { kind, source }),
+        }
+    }
+
+    pub fn new_with_source<E: Into<BoxError>>(kind: ExecutionErrorKind, source: E) -> Self {
+        Self::new(kind, Some(source.into()))
+    }
+
+    pub fn from_kind(kind: ExecutionErrorKind) -> Self {
+        Self::new(kind, None)
+    }
+
+    pub fn kind(&self) -> &ExecutionErrorKind {
+        &self.inner.kind
+    }
+
+    pub fn to_execution_status(&self) -> ExecutionFailureStatus {
+        match self.kind() {
+            ExecutionErrorKind::InsufficientGas => ExecutionFailureStatus::InsufficientGas,
+            ExecutionErrorKind::TransferUnowned
+            | ExecutionErrorKind::TransferNonCoin
+            | ExecutionErrorKind::TransferInsufficientBalance
+            | ExecutionErrorKind::InvalidTransactionUpdate
+            | ExecutionErrorKind::ObjectNotFound
+            | ExecutionErrorKind::DeleteObjectOwnedObject
+            | ExecutionErrorKind::FunctionNotFound
+            | ExecutionErrorKind::ModuleNotFound
+            | ExecutionErrorKind::InvalidFunctionSignature
+            | ExecutionErrorKind::InvalidFunctionVisibility
+            | ExecutionErrorKind::InvalidNonEntryFunction
+            | ExecutionErrorKind::ExecutionInvariantViolation
+            | ExecutionErrorKind::TypeError
+            | ExecutionErrorKind::CircularObjectOwnership
+            | ExecutionErrorKind::MissingObjectOwner
+            | ExecutionErrorKind::InvalidSharedChildUse
+            | ExecutionErrorKind::ModulePublishFailure
+            | ExecutionErrorKind::ModuleVerificationFailure => {
+                ExecutionFailureStatus::MiscellaneousError
+            }
+            ExecutionErrorKind::VmError => {
+                let source = if let Some(source) = self.source() {
```

Comment:
I have a personal crusade here, so take it or leave it :) 
I really dislike `if-let-else` as it just feels like a worse version of `match` to me. (where `if-let` is great for not having to write `match e { p => ..., _ => () }`, I don't feel like anything is gained from `if-let` with an `else`)

Hopefully `let-else` lands some day and I can stop my crusade for cases like this one at lease.